### ISSUE TITLE
Should we return unique_ptr or shared_ptr from BATSMsgFactory::createMsg()

### ISSUE THREAD

**Issue by bigfatwhale** [bigfatwhale/bats_pitch_parser#3]
Is it overkill to return shared_ptr? Would it be more performant to return unique_ptr instead? Actually performance difference might not be that big though, but worth pondering about it from a design stand-point.

**Comment by bigfatwhale** [comment]
RVO has some technicalities as well, and might not always work. See [here]( https://www.ibm.com/developerworks/community/blogs/5894415f-be62-4bc0-81c5-3956e82276f3/entry/RVO_V_S_std_move?lang=en).

There's also another issue of slicing if we switch to copy because for the factory method we are dynamically constructing one of the derived class inside the createMsg() and then returning it as a base class pointer. That works fine. 

If instead we switch to, 
``` C++
BATSMessageBase BATSMsgFactory::createMsg()
```
then the derived class object instance (say BATSAuctionSummaryMsg) will be sliced when it's returned as a BATSMessageBase.

Is my understanding correct or maybe I'm not getting the full picture...

**Comment by bigfatwhale** [comment]
Another thing that I not quite sure about is whether it will then also complicate the exported Python API stuff, 

```python
void BATSMessageBase::export_to_python()
{
    //boost::python::register_ptr_to_python<std::shared_ptr<BATSMessageBase>>();
    boost::python::class_<BATSMessageBase,
                          std::shared_ptr<BATSMessageBase>,
                          boost::noncopyable >("BATSMessageBase")
            .def(boost::python::init<>())
            .def(boost::python::init<int, char>())
            .def_readwrite("timestamp", &BATSMessageBase::m_timestamp)
            .def_readwrite("msgtype", &BATSMessageBase::m_msgtype)
            .def("__repr__", &BATSMessageBase::repr);
}
```
The ```shared_ptr<BATSMessageBase>``` works nicely out of the box when exporting the ```BATSPitchMsgParser::parse_msg``` function to Pyhon.

**Comment by bigfatwhale** [comment]
I was referring to something like the code below when I was commenting on the issue of slicing in the previous comment, 
```c++
class A
{
public:
    int a;
};

class B : public A
{
    int b;
};

A foo() // factory method
{
    B b;
    return b;
}

int main() {
    auto x = foo();
    cout << x.a << endl; // x won't have the field b
}
```

**Comment by wlyeow** [comment]
You're right. Don't know why I completely missed the inherited structs. Maybe it's because I'm trying too hard to avoid writing like Java in C++.

**Comment by wlyeow** [comment]
If sticking to polymorphism, the ideal answer should be `unique_ptr` since we don't want the parser to leak. But I don't know how supported this is in Spirits. Something more to research...

**Comment by bigfatwhale** [comment]
Yar i think at various places the recommended guidelines for factory methods is to return unique_ptr.

I think you're right, maybe we should abstract the python part out, 

1. Let the C++ users use the unique_ptr version of the factory method ```unique_ptr<BATSMessageBase>
BATSMsgFactory::createMsg()```. 
2. For the part we export to Python, we can convert the unique_ptr to shared_ptr.

**Comment by bigfatwhale** [comment]
I tried this out and not much difference in performance. What's left to consider is whether it makes more sense to be unique_ptr instead of shared_ptr from a software-engineering standpoint.

Comments?
